### PR TITLE
Improvements to Documentation

### PR DESCRIPTION
**Pull request by KyeMaloy97**
Improved the documentation and raised the coverage to 100%.

## Description
Updated the docs for the classes, variables, protocols and functions and also added usage examples for everything.

## Motivation and Context
The old docs were good but didn't provide usage examples and also needed updating as they were sometimes short or to succinct. 

## How Has This Been Tested?
Only doc changes so no code has been edited. The web pages load correctly in my browser.

## Checklist:
- [X] If applicable, I have updated the documentation accordingly.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,11 +1,53 @@
...
+<img src="http://swift-at-ibm-slack.mybluemix.net/badge.svg" alt="Slack Status">
+</a>
+</p>
 
-The KituraNet module contains logic for sending and receiving HTTP requests. It also contains the structure for listening on a port and sending requests to a delegate for processing.
+# Kitura-Net
+
+The Kitura-net module contains logic for sending and receiving HTTP requests. It also contains the structure for listening on a port and sending requests to a delegate for processing. It can be used to create HTTP/CGI Servers on specific ports, and provides HTTP functionality.
+
+If you require Routing, Templates or Middleware functionality, please see the [Kitura](https://github.com/IBM-Swift/Kitura) project.
```
Comment (helenmasters) [inline README.md]: We need to be clearer about when and what we expect end users to be using Kitura-net for. To that end I think we need to have a longer explanation here - "We expect most of our users to require higher level concepts such as routing, templates and middleware, these are not provided in Kitura-net, if you want to use those facilities you should be coding at the Kitura level, for this please see the [Kitura](https://github.com/IBM-Swift/Kitura) project."

Comment (helenmasters) [reply README.md]: We may even want to say that Kitura uses Kitura-net under the covers...

File: README.md [hunk 2]
```diff
@@ -1,11 +1,53 @@
...
+
+1. Clone this repository.
+
+`$ git clone https://github.com/IBM-Swift/Kitura-net`
+
+2. Build and run tests.
+
+`$ swift test`
+
+You can find more info on contributing to Kitura in our [contributing guidelines](https://github.com/IBM-Swift/Kitura/blob/master/.github/CONTRIBUTING.md).
```
Comment (helenmasters) [inline README.md]: Please can you add a Community section below this, with the following text in:

We love to talk server-side Swift, and Kitura. Join our [Slack](http://swift-at-ibm-slack.mybluemix.net/) to meet the team!

File: Sources/KituraNet/BufferList.swift [hunk 3]
```diff
@@ -127,16 +195,29 @@ public class BufferList {
         
     }
     
-    /// Resets the buffer to zero length and the beginning position
+    /**
+     Resets the buffer to zero length and the beginning position.
```
Comment (helenmasters) [inline Sources/KituraNet/BufferList.swift]: This sentence doesn't make sense to me, specifically the words "and the beginning position".  How about "Resets the buffer to the beginning position and the buffer length to zero."

File: Sources/KituraNet/BufferList.swift [hunk 4]
```diff
@@ -112,12 +174,18 @@ public class BufferList {
...
     }
     
-    /// Fill an NSMutableData with data from the buffer
-    ///
-    /// - Parameter data: The NSMutableData object to fill from data in the buffer
-    ///
-    /// - Returns: The number of bytes actually copied from the buffer. It will be equal
-    ///           to the number of bytes left in the buffer.
+    /**
+     Fill a NSMutableData with data from the buffer.
```
Comment (helenmasters) [inline Sources/KituraNet/BufferList.swift]: Fill a `NSMutableData` object with data from the buffer.

File: Sources/KituraNet/BufferList.swift [hunk 5]
```diff
@@ -18,74 +18,129 @@ import Foundation
...
+    ### Usage Example: ###
+    ````swift
+    //Create a BufferList object.
+    var writeBuffer = BufferList()
+ 
+    //Create a new constant of type Int of the number of bytes copied from the buffer.
+    let count = writeBuffer.fill(buffer: UnsafeMutableRawPointer(buf).assumingMemoryBound(to: UInt8.self), length: size)
+ 
+    //Append some data to the initialized writeBuffer object.
+    writeBuffer.append(bytes + written, length: length - written)
```
Comment (helenmasters) [inline Sources/KituraNet/BufferList.swift]: Please can we have a working example here as this contains too many undefined variables to be really helpful.

File: Sources/KituraNet/BufferList.swift [hunk 6]
```diff
@@ -26,13 +26,17 @@ import Foundation
     //Create a BufferList object.
```
Comment (helenmasters) [inline Sources/KituraNet/BufferList.swift]: I've been playing with BufferList as I couldn't get the new example to work, try this as our code example:
```swift
var writeBuffer = BufferList()
            
let firstArrayOfNumbers: [UInt8] = [1,2,3,4]
            
// Append a set of data to the `writeBuffer` object
writeBuffer.append(bytes: UnsafePointer<UInt8>(firstArrayOfNumbers), length: MemoryLayout<UInt8>.size * firstArrayOfNumbers.count)
            
// Number of bytes stored in the `writeBuffer` object
print(writeBuffer.count)
// Prints "4"
            
let secondArrayOfNumbers: [UInt8] = [5,6,7,8]
            
// Append a second set of data to the `writeBuffer` object
writeBuffer.append(bytes: UnsafePointer<UInt8>(secondArrayOfNumbers), length: MemoryLayout<UInt8>.size * secondArrayOfNumbers.count)
            
print(writeBuffer.count)
// Prints "8"
            
let finalArrayOfNumbers: [UInt8] = [0,0,0,0,0,0,0,0,9,10]
            
// Fill the destination buffer `finalArrayOfNumbers` with the data from `writeBuffer`
let count = writeBuffer.fill(buffer: UnsafeMutableRawPointer(mutating: finalArrayOfNumbers).assumingMemoryBound(to: UInt8.self), length: ((MemoryLayout<UInt8>.size) * finalArrayOfNumbers.count))
            
print("count = \(count), buffer is = \(finalArrayOfNumbers)" )
// Prints "count = 8, buffer is = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]"
```

File: Sources/KituraNet/BufferList.swift [hunk 7]
```diff
@@ -18,74 +18,129 @@ import Foundation
...
 // MARK: BufferList 
 
-/// This class provides an implementation of a buffer that can be added to and
-/// "taken" from in chunks. Data is always added to the end of the buffer and
-/// data is "taken" from the buffer from the beginning towards the end. A pointer
-/// is maintained as where the next chunk of data will be "taken" from. This
-/// pointer can be reset, to enable the data in the buffer to be fetched from the
-/// beginning again.
+    /**
+    This class provides an implementation of a buffer that can be added to and "taken" from in chunks. Data is always added to the end of the buffer and data is "taken" from the buffer from the beginning towards the end. A pointer is maintained as where the next chunk of data will be "taken" from. This pointer can be reset, to enable the data in the buffer to be fetched from the beginning again.
```
Comment (helenmasters) [inline Sources/KituraNet/BufferList.swift]: Please remove the "" from around the word taken as that's not adding anything helpful here.

I think we need the following wording (or similar) above the example I've added below...
"When you create a `BufferList` instance it is empty. You can then append data to your `BufferList` instance, in our case called 'writeBuffer'. In the example below we make two separate appends. When 'writeBuffer' contains the data which you wish to write out you can use `BufferList.fill(...)` to write out the data from the buffer to your chosen location, in this case 'finalArrayOfNumbers'.

File: Sources/KituraNet/BufferList.swift [hunk 8]
```diff
@@ -18,74 +18,129 @@ import Foundation
 
 // MARK: BufferList 
 
-/// This class provides an implementation of a buffer that can be added to and
-/// "taken" from in chunks. Data is always added to the end of the buffer and
-/// data is "taken" from the buffer from the beginning towards the end. A pointer
-/// is maintained as where the next chunk of data will be "taken" from. This
-/// pointer can be reset, to enable the data in the buffer to be fetched from the
-/// beginning again.
```
Comment (helenmasters) [inline Sources/KituraNet/BufferList.swift]: Why don't we reword this as "This class provides an implementation of a buffer than can be added to and taken from in chunks.  Data is always added to the end of the buffer (using `BuffedList.append(...)` and taken out of the buffer (using `BufferedList.fill(...)`) from the beginning towards the end.  The location indicating where the next chunk of data will be taken from is maintained, this location can then be reset to enable data to be taken out of the buffer from the beginning again."

File: Sources/KituraNet/BufferList.swift [hunk 9]
```diff
@@ -18,74 +18,129 @@ import Foundation
...
+     Byte offset inside of internal storage buffer
+    */
     private var byteIndex = 0
     
-    // MARK: -- Public 
+    // MARK: Public
     
-    /// Get the number of bytes stored in the BufferList
+    /**
+     Get the number of bytes stored in the BufferList.
```
Comment (helenmasters) [inline Sources/KituraNet/BufferList.swift]: Get the number of bytes stored in the `BufferList`.

File: Sources/KituraNet/BufferList.swift [hunk 10]
```diff
@@ -18,74 +18,129 @@ import Foundation
...
+    */
     public var count: Int {
         return localData.length
     }
     
-    /// Read the data in the BufferList
+    /**
+     Read the data in the BufferList.
```
Comment (helenmasters) [inline Sources/KituraNet/BufferList.swift]: Read the data from the `BufferList`.

File: Sources/KituraNet/BufferList.swift [hunk 11]
```diff
@@ -18,74 +18,129 @@ import Foundation
...
     
-    /// Initializes a BufferList instance
-    ///
-    /// - Returns: a BufferList instance
+    /**
+     Creates a BufferList instance to store bytes to be written.
+     
+     ### Usage Example: ###
+     ````swift
+     var writeBuffers = BufferList()
```
Comment (helenmasters) [inline Sources/KituraNet/BufferList.swift]: ```var writeBuffer = BufferList()``` I'd rather the variable had a singular name not plural.

File: Sources/KituraNet/BufferList.swift [hunk 12]
```diff
@@ -18,74 +18,129 @@ import Foundation
...
+    */
     public var data: Data {
         return Data(bytes: localData.bytes, count: localData.length)
     }
     
-    /// Initializes a BufferList instance
-    ///
-    /// - Returns: a BufferList instance
+    /**
+     Creates a BufferList instance to store bytes to be written.
```
Comment (helenmasters) [inline Sources/KituraNet/BufferList.swift]: Creates a `BufferList` instance in which to store bytes to be written.

File: Sources/KituraNet/BufferList.swift [hunk 13]
```diff
@@ -18,74 +18,129 @@ import Foundation
...
-    /// Append bytes to the buffer
-    ///
-    /// Parameter bytes: The pointer to the bytes
-    /// Parameter length: The number of bytes to append
+    /**
+     Append bytes to the buffer.
+     
+     ### Usage Example: ###
+     ````swift
+     writeBuffer.append(bytes + written, length: length - written)
```
Comment (helenmasters) [inline Sources/KituraNet/BufferList.swift]: I'd prefer the following example:
```swift
var writeBuffer = BufferList()
let firstArrayOfNumbers: [UInt8] = [1,2,3,4]
writeBuffer.append(bytes: UnsafePointer<UInt8>(firstArrayOfNumbers), length: MemoryLayout<UInt8>.size * firstArrayOfNumbers.count)
```

File: Sources/KituraNet/BufferList.swift [hunk 14]
```diff
@@ -18,74 +18,129 @@ import Foundation
...
     public func append(bytes: UnsafePointer<UInt8>, length: Int) {
         localData.append(bytes, length: length)
     }
     
-    /// Append data into BufferList 
-    /// 
-    /// Parameter data: The data to append
+
+    /**
+     Append data into the BufferList.
```
Comment (helenmasters) [inline Sources/KituraNet/BufferList.swift]: Append data into the `BufferList` instance.

File: Sources/KituraNet/BufferList.swift [hunk 15]
```diff
@@ -18,74 +18,129 @@ import Foundation
...
-    ///           of the memory area provided.
+    /**
+     Fill memory with data from the buffer.
+     
+     ### Usage Example: ###
+     ````swift
+     let count = writeBuffer.fill(buffer: UnsafeMutableRawPointer(buf).assumingMemoryBound(to: UInt8.self), length: size)
+     ````
+     
+     - Parameter buffer: A NSMutablePointer to the beginning of the memory to be filled.
```
Comment (helenmasters) [inline Sources/KituraNet/BufferList.swift]: `` around NSMutablePointer please!

File: Sources/KituraNet/BufferList.swift [hunk 16]
```diff
@@ -96,13 +151,20 @@ public class BufferList {
...
     }
-    
-    /// Fill a Data struct with data from the buffer
-    ///
-    /// - Parameter data: The Data struct to fill from data in the buffer
-    ///
-    /// - Returns: The number of bytes actually copied from the buffer. It will be equal
-    ///           to the number of bytes left in the buffer.
+
+    /**
```
Comment (helenmasters) [inline Sources/KituraNet/BufferList.swift]: Please can you make Structure have a lowercase s in all the places it's referenced below.

File: Sources/KituraNet/BufferList.swift [hunk 17]
```diff
@@ -96,13 +151,20 @@ public class BufferList {
...
+     Fill a Data Structure with data from the buffer.
+     
+     ### Usage Example: ###
+     ````swift
+     let count = writeBuffer.fill(data: &data)
+     ````
+     
+     - Parameter data: The Data Structure to fill from the data in the buffer.
```
Comment (helenmasters) [inline Sources/KituraNet/BufferList.swift]: The `Data` structure to fill with data from the buffer.

File: Sources/KituraNet/BufferList.swift [hunk 18]
```diff
@@ -112,12 +174,18 @@ public class BufferList {
...
-    /// Fill an NSMutableData with data from the buffer
-    ///
-    /// - Parameter data: The NSMutableData object to fill from data in the buffer
-    ///
-    /// - Returns: The number of bytes actually copied from the buffer. It will be equal
-    ///           to the number of bytes left in the buffer.
+    /**
+     Fill a NSMutableData with data from the buffer.
+     
+     - Parameter data: The NSMutableData object to fill from the data in the buffer.
```
Comment (helenmasters) [inline Sources/KituraNet/BufferList.swift]: Needs single quotes around NSMutableData.

File: Sources/KituraNet/BufferList.swift [hunk 19]
```diff
@@ -127,16 +195,29 @@ public class BufferList {
...
         
         localData.length = 0
         byteIndex = 0
         
     }
     
-    /// Sets the buffer back to the beginning position, the next fill call will take data
-    /// from the beginning of the buffer.
+    /**
+     Sets the buffer back to the beginning position. The next fill call will take data from the beginning of the buffer.
```
Comment (helenmasters) [inline Sources/KituraNet/BufferList.swift]: `BufferList.fill(...)` instead of fill

File: Sources/KituraNet/BufferList.swift [hunk 20]
```diff
@@ -18,74 +18,129 @@ import Foundation
...
         return fill(buffer: UnsafeMutablePointer(mutating: array), length: array.count)
     }
     
-    /// Fill memory with data from the buffer
-    ///
-    /// - Parameter buffer: NSMutablePointer to the beginning of the memory to be filled
-    /// - Parameter length: The number of bytes to fill
-    ///
-    /// - Returns: The number of bytes actually copied from the buffer. It will be the
-    ///           lessor of the number of bytes left in the buffer and the length
```
Comment (helenmasters) [inline Sources/KituraNet/BufferList.swift]: typo - lesser not lessor!
Also the return sentence doesn't make sense to me surely... "It will be the lesser of the number of bytes in the buffer we are copying data from and the length of the memory area provided."

File: Sources/KituraNet/BufferList.swift [hunk 21]
```diff
@@ -112,12 +174,18 @@ public class BufferList {
...
-    /// - Parameter data: The NSMutableData object to fill from data in the buffer
-    ///
-    /// - Returns: The number of bytes actually copied from the buffer. It will be equal
-    ///           to the number of bytes left in the buffer.
+    /**
+     Fill a NSMutableData with data from the buffer.
+     
+     - Parameter data: The NSMutableData object to fill from the data in the buffer.
+     
+     - Returns: The number of bytes actually copied from the buffer. It will be equal to the number of bytes left in the buffer.
```
Comment (helenmasters) [inline Sources/KituraNet/BufferList.swift]: Can't we delete the sentence "It will be equal to the number of bytes left in the buffer."?

File: Sources/KituraNet/ClientRequest.swift [hunk 22]
```diff
@@ -21,35 +21,97 @@ import Socket
 import Foundation
 
 // MARK: ClientRequest
-
-/// This class provides a set of low level APIs for issuing HTTP requests to another server.
+/**
+This class provides a set of low level APIs for issuing HTTP requests to another server. A new instance of the request can be created, along with options if the user would like to specify certain parameteres such as HTTP headers, HTTP methods, host names, and SSL credentials. Data and Strings can also be written to it, and URLs can be parsed.
```
Comment (helenmasters) [inline Sources/KituraNet/ClientRequest.swift]: typo - parameteres should be parameters

Comment (helenmasters) [reply Sources/KituraNet/ClientRequest.swift]: "Data and Strings can also be written to it, and URLs can be parsed." This sentence doesn't make sense to me, do we need it/what are we trying to say here?

File: README.md (hunk 1 above)
Comment (KyeMaloy97) [reply README.md]: ✅

File: Sources/KituraNet/BufferList.swift (hunk 3 above)
Comment (KyeMaloy97) [reply Sources/KituraNet/BufferList.swift]: ✅

File: Sources/KituraNet/BufferList.swift (hunk 4 above)
Comment (KyeMaloy97) [reply Sources/KituraNet/BufferList.swift]: ✅

File: Sources/KituraNet/BufferList.swift (hunk 11 above)
Comment (KyeMaloy97) [reply Sources/KituraNet/BufferList.swift]: ✅

File: Sources/KituraNet/BufferList.swift (hunk 10 above)
Comment (KyeMaloy97) [reply Sources/KituraNet/BufferList.swift]: ✅

File: Sources/KituraNet/BufferList.swift (hunk 9 above)
Comment (KyeMaloy97) [reply Sources/KituraNet/BufferList.swift]: ✅

File: Sources/KituraNet/BufferList.swift (hunk 12 above)
Comment (KyeMaloy97) [reply Sources/KituraNet/BufferList.swift]: ✅

File: Sources/KituraNet/BufferList.swift [hunk 23]
```diff
@@ -18,74 +18,129 @@ import Foundation
...
-    /// - Returns: a BufferList instance
+    /**
+     Creates a BufferList instance to store bytes to be written.
+     
+     ### Usage Example: ###
+     ````swift
+     var writeBuffers = BufferList()
+     ````
+     
+     - Returns: A BufferList instance.
```
[the remark this answers is not in the document]
Comment (KyeMaloy97) [reply Sources/KituraNet/BufferList.swift]: ✅

File: Sources/KituraNet/BufferList.swift (hunk 14 above)
Comment (KyeMaloy97) [reply Sources/KituraNet/BufferList.swift]: ✅

File: Sources/KituraNet/BufferList.swift (hunk 15 above)
Comment (KyeMaloy97) [reply Sources/KituraNet/BufferList.swift]: ✅

File: Sources/KituraNet/BufferList.swift (hunk 17 above)
Comment (KyeMaloy97) [reply Sources/KituraNet/BufferList.swift]: I haven't put Data in marks because it doesn't mean a Swift Data object, just any computer data structure.

File: Sources/KituraNet/BufferList.swift (hunk 18 above)
Comment (KyeMaloy97) [reply Sources/KituraNet/BufferList.swift]: ✅

File: Sources/KituraNet/BufferList.swift (hunk 19 above)
Comment (KyeMaloy97) [reply Sources/KituraNet/BufferList.swift]: ✅

File: Sources/KituraNet/BufferList.swift (hunk 20 above)
Comment (KyeMaloy97) [reply Sources/KituraNet/BufferList.swift]: ✅

File: Sources/KituraNet/BufferList.swift [hunk 24]
```diff
@@ -18,74 +18,129 @@ import Foundation
...
+     
+     ### Usage Example: ###
+     ````swift
+     let count = writeBuffer.fill(buffer: UnsafeMutableRawPointer(buf).assumingMemoryBound(to: UInt8.self), length: size)
+     ````
+     
+     - Parameter buffer: A NSMutablePointer to the beginning of the memory to be filled.
+     - Parameter length: The number of bytes to fill.
+     
+     - Returns: The number of bytes actually copied from the buffer. It will be the lesser of the number of bytes left in the buffer and the length of the memory area provided.
```
[the remark this answers is not in the document]
Comment (KyeMaloy97) [reply Sources/KituraNet/BufferList.swift]: ✅

File: Sources/KituraNet/BufferList.swift [hunk 25]
```diff
@@ -96,13 +151,20 @@ public class BufferList {
...
+     Fill a Data Structure with data from the buffer.
+     
+     ### Usage Example: ###
+     ````swift
+     let count = writeBuffer.fill(data: &data)
+     ````
+     
+     - Parameter data: The Data Structure to fill from the data in the buffer.
+     
+     - Returns: The number of bytes actually copied from the buffer. It will be equal to the number of bytes left in the buffer.
```
[the remark this answers is not in the document]
Comment (KyeMaloy97) [reply Sources/KituraNet/BufferList.swift]: ✅

File: Sources/KituraNet/BufferList.swift (hunk 21 above)
Comment (KyeMaloy97) [reply Sources/KituraNet/BufferList.swift]: ✅

File: Sources/KituraNet/ClientRequest.swift (hunk 22 above)
Comment (KyeMaloy97) [reply Sources/KituraNet/ClientRequest.swift]: ✅

File: Sources/KituraNet/BufferList.swift (hunk 16 above)
Comment (KyeMaloy97) [reply Sources/KituraNet/BufferList.swift]: ✅

File: Sources/KituraNet/BufferList.swift (hunk 13 above)
Comment (KyeMaloy97) [reply Sources/KituraNet/BufferList.swift]: ✅

File: Sources/KituraNet/BufferList.swift (hunk 6 above)
Comment (KyeMaloy97) [reply Sources/KituraNet/BufferList.swift]: ✅

File: Sources/KituraNet/BufferList.swift (hunk 8 above)
Comment (KyeMaloy97) [reply Sources/KituraNet/BufferList.swift]: ✅

**Comment by KyeMaloy97** [comment]
Opening in lots of smaller PRs as this PR is too large to be reviewed properly.
